test/functional/p2p_leak.py

```diff
@@ -106,39 +95,51 @@ def set_test_params(self):
         self.num_nodes = 1
 
     def run_test(self):
-        no_version_disconnect_node = self.nodes[0].add_p2p_connection(
-            CNodeNoVersionMisbehavior(), send_version=False, wait_for_verack=False)
-        no_version_idlenode = self.nodes[0].add_p2p_connection(CNodeNoVersionIdle(), send_version=False, wait_for_verack=False)
-        no_verack_idlenode = self.nodes[0].add_p2p_connection(CNodeNoVerackIdle(), wait_for_verack=False)
+        # Peer that never sends a version. We will send a bunch of messages
+        # from this peer anyway and verify eventual disconnection.
+        no_version_disconnect_peer = self.nodes[0].add_p2p_connection(
+            LazyPeer(), send_version=False, wait_for_verack=False)
+
+        # Another peer that never sends a version, nor any other messages. It shouldn't receive anything from the node.
+        no_version_idle_peer = self.nodes[0].add_p2p_connection(LazyPeer(), send_version=False, wait_for_verack=False)
+
+        # Peer that sends a version but not a verack.
+        no_verack_idle_peer = self.nodes[0].add_p2p_connection(NoVerackIdlePeer(), wait_for_verack=False)
 
-        # Send enough veracks without a message to reach the peer discouragement
-        # threshold. This should get us disconnected.
+        # Send enough ping messages (any non-version message will do) prior to sending
+        # version to reach the peer discouragement threshold. This should get us disconnected.
         for _ in range(DISCOURAGEMENT_THRESHOLD):
-            no_version_disconnect_node.send_message(msg_verack())
+            no_version_disconnect_peer.send_message(msg_ping())
 
-        # Wait until we got the verack in response to the version. Though, don't wait for the other node to receive the
+        # Wait until we got the verack in response to the version. Though, don't wait for the node to receive the
         # verack, since we never sent one
-        no_verack_idlenode.wait_for_verack()
+        no_verack_idle_peer.wait_for_verack()
+
+        wait_until(lambda: no_version_disconnect_peer.ever_connected, timeout=10, lock=mininode_lock)
+
+        # Wait until the node receives at least as many pings from no_version_disconnect_peer (peer 0)
+        # as we sent in the test
+        wait_until(lambda: self.nodes[0].getpeerinfo()[0]['bytesrecv_per_msg'].get('ping', 0) >=
```

Comment:
Removed this assertion. Thank you for the hints. I learnt that occasionally the misbehaving peer is removed before this assertion even runs, and so the assertion fails. Since this peer does nothing else except send a 100 pings, i think it suffices to check they are no long connected.